### ISSUE TITLE
[css-values] Conditional syntax for `if()` 

### ISSUE THREAD

**Issue by LeaVerou** [w3c/csswg-drafts#8446]
Over the years there have been several issues and proposals on an inline `if()` function:

- #5009
- https://github.com/w3c/csswg-drafts/issues/5624
- My old [css-conditional-values unofficial draft](https://drafts.csswg.org/css-conditional-values/#if)

My understanding is that we generally have consensus that we need this, but the devil is in the details. 
Now that style container queries are [shipping](https://groups.google.com/a/chromium.org/g/blink-dev/c/ACL23q_nbK0/m/PaNJ81_qDAAJ), inline `if()` is the only missing piece for being able to use higher level custom properties to customize web components, and it would be nice if they could ship somewhat close to each other.

It seems that one of the primary blockers for `if()` is syntax, and especially the syntax of the condition itself.

### Condition

I see two directions for this:
a) We adopt the style container query syntax wholesale, and accept `<style-query>` as the condition. 
b) We define a new syntax in terms of comparing *values*, similar to what [css-conditional-values unofficial draft](https://drafts.csswg.org/css-conditional-values/#if) was trying to do.

The upside of (a) is consistency and learnability, it's much easier for authors to learn one conditional syntax for both places.
The upside of (b) is that it's strictly more powerful. Since `if()` is used on a specific property, resolving all values is possible, so the constraints of the container query syntax don't really apply. Also, there are plans to extend `<style-query>` to arbitrary properties, which may not be desirable for `if()`.

### Other syntax considerations

What should we use to separate the condition from the iftrue and iffalse values? In my original design I used commas, but this means we cannot use any values that *contain* commas, so the direction [`mix()`](https://drafts.csswg.org/css-values/#interpolate), [`first-supported()`](https://github.com/w3c/csswg-drafts/issues/5055) have been going with semicolons as the separator seems preferable.

cc @tabatkins @mirisuzanne

**Comment by muratcorlu** [comment]
> inline `if()` is the only missing piece for being able to use higher level custom properties to customize web components, 

Once we have style container queries with [extended style query syntax](https://github.com/w3c/csswg-drafts/issues/7068), do we still need to have `if()` for customizing web components with CSS variables? Can you give an example that can not be done with style queries? Or do we want to give another practical option to the developers?

Web components always have a container and variables will be always applied to the container element. As far as I see, `if()` is only helpful if you don't want to rely on a container, or as an "alternative" to style queries. Am I missing a point here?

**Comment by mirisuzanne** [comment]
The inline `if()` would have advantages in several cases:

- For cases where an author needs different conditions for individual properties, the container at-rule becomes very repetitive. 
- All the use-cases that could be applied to web components with a host and shadow-dom (where a host is built-in) can also be applied to 'components' built with normal elements and no shadow dom. In those cases, there is no guarantee of a container. There's no reason for us to consider shadow-dom the primary use-case for these features.
- As noted by @LeaVerou above, in the context of a property we can allow comparing eg numbers with percentages.

   > Since if() is used on a specific property, resolving all values is possible, so the constraints of the container query syntax don't really apply.

**Comment by LeaVerou** [comment]
To add to what @mirisuzanne said above, even with web components, you often want to style the element itself. Not everything is done in shadow DOM.

**Comment by benface** [comment]
Would this proposed `if()` function allow for something like this?

```css
display: if(--hidden: 1, none, flex);
```

**Comment by LeaVerou** [comment]
@benface yes, except the syntax would look more like `if(style(--hidden: 1), none, flex)`.

**Comment by LeaVerou** [comment]
Yes.

**Comment by tabatkins** [comment]
Closing this, as it's now supported by the `if()` spec via a `style()` query. The example from a few comments up would be written `if(style(hidden: 1): none; else: flex)`.
